File: packages/ld-adapter/src/resolve/rule-resolver.ts
import type { FlagState, RuleRef } from '@ff-promo/contracts';
import { UnresolvedRuleError } from '../errors/ld-adapter-error.js';

export function resolveRuleId(
	flagState: FlagState,
	environmentKey: string,
	ref: RuleRef,
): string {
	const env = flagState.environments[environmentKey];
	if (!env) {
		throw new UnresolvedRuleError(
			`Environment not found for rule resolution: ${environmentKey}`,
			{ environmentKey, flagKey: flagState.flagKey, ref },
		);
	}

	const matches = env.rules.filter((rule) => {
		switch (ref.by) {
			case 'id':
				return rule.id === ref.id;
			case 'description':
				return rule.description === ref.description;
			default:
				return false;
		}
	});

	if (matches.length !== 1) {
		throw new UnresolvedRuleError(
			`Expected exactly one rule match in ${environmentKey}, found ${matches.length}`,
			{ environmentKey, ref, flagKey: flagState.flagKey },
		);
	}

	const match = matches[0];
	if (!match?.id) {
		throw new UnresolvedRuleError('Matched rule is missing id', {
			environmentKey,
			ref,
		});
	}

	return match.id;
}

File: kotlin/modules/ld-adapter/src/main/kotlin/com/ffpromo/ldadapter/resolve/RuleResolver.kt
package com.ffpromo.ldadapter.resolve

import com.ffpromo.contracts.FlagState
import com.ffpromo.contracts.RuleRef
import com.ffpromo.ldadapter.errors.UnresolvedRuleError

fun resolveRuleId(
    flagState: FlagState,
    environmentKey: String,
    ref: RuleRef,
): String {
    val env = flagState.environments[environmentKey]
        ?: throw UnresolvedRuleError(
            message = "Environment not found for rule resolution: $environmentKey",
            context = mapOf(
                "environmentKey" to environmentKey,
                "flagKey" to flagState.flagKey,
                "ref" to ref,
            ),
        )

    val matches = env.rules.filter { rule ->
        when (ref) {
            is RuleRef.ById -> rule.id == ref.id
            is RuleRef.ByDescription -> rule.description == ref.description
        }
    }

    if (matches.size != 1) {
        throw UnresolvedRuleError(
            message = "Expected exactly one rule match in $environmentKey, found ${matches.size}",
            context = mapOf(
                "environmentKey" to environmentKey,
                "ref" to ref,
                "flagKey" to flagState.flagKey,
            ),
        )
    }

    val match = matches.first()
    if (match.id.isEmpty()) {
        throw UnresolvedRuleError(
            message = "Matched rule is missing id",
            context = mapOf("environmentKey" to environmentKey, "ref" to ref),
        )
    }

    return match.id
}
